content/blog/2017/2017-02-23-declarative-saucelabs-xunit.adoc

```diff
@@ -0,0 +1,400 @@
+
+== Elevating Settings using `environment`
+
+Each time we've moved a project from Scripted Pipeline to Declarative,
+we've found the cleaner format of Declarative Pipeline highlights the less
+clear parts of the existing Pipeline.
+In this case, the first thing that jumps out at me is that the parameters of the
+Saucelabs and Nightwatch execution are hardcoded and buried down in the middle of our Pipeline.
+This is a relatively short Pipeline, so it isn't terribly hard to find them,
+but as this pipeline grows and changes it would be better if those values were kept separate.
+In Scripted, we'd have defined some variables,
+but Declarative doesn't allow us to define variables in the usual Groovy sense.
+
+The `environment` directive let's us set some environment variables
+and use them later in our pipeline.
+As you'd expect, the `environment` directive is just a set of name-value pairs.
+Environment variables are accessible in Pipeline via `env.variableName` (or just `variableName`)
+and in shell scripts as standard environment variables, typically `$variableName`.
+
+Let's move the list of browsers, the test filter, and the sauce credential string to environment variables.
+
+.Jenkinsfile
+[source, groovy]
+----
+    environment {
+        saucelabsCredentialId = 'f0a6b8ad-ce30-4cba-bf9a-95afbc470a8a'
+        sauceTestFilter = 'tests/guineaPig.js'
+        platformConfigs = 'chrome,firefox,ie,edge'
+    }
+    stages {
+        /* ... unchanged ... */
+        stage ("Test") {
+            steps {
+                // Add sauce credentials
+                sauce(saucelabsCredentialId) {
+                    // Start sauce connect
+                    sauceconnect() {
+                        // Run selenium tests using Nightwatch.js
+                        // Ignore error codes. The junit publisher will cover setting build status.
+                        sh "./node_modules/.bin/nightwatch -e ${env.platformConfigs} --test ${env.sauceTestFilter} || true" // <1>
+                    }
+                }
+            }
+            post { /* ... unchanged ... */ }
+        }
+    }
+}
+----
+<1>  This double-quoted string causes Groovy to replace the variables with their
+literal values before passing to `sh`.
+This could also be written using singe-quotes:
+`sh './node_modules/.bin/nightwatch -e $platformConfigs --test $sauceTestFilter || true'`.
+With a single quoted string, the string is passed as written to the shell,
+and then *the shell* does the variable substitution.
+
+== Moving Complex Code to Shared Libraries
+
+Now that we have settings separated from the code, we can do some code clean up.
+Unlike the previous post, we don't have any repeating code,
+but we do have some distractions.
+The nesting of `sauce`, `sauceconnect`, and `sh nightwatch` seems excessive,
+and that xUnit `step` is a bit ugly as well.
+Let's move those into our shared library as custom steps with parameters.
```

Comment:
Please link "shared library" as "[Shared Library](https://jenkins.io/doc/book/pipeline/shared-libraries)" to the relevant handbook section